### ISSUE TITLE
ONNX output doesn't match PyTorch when export is run on M1 model.

### ISSUE THREAD

**Issue by NicolasNierenbergPersyst** [onnx/onnx#4435]
In the specific case of a PyTorch model with a Conv2d and a BatchNorm2d if export is run on an M1 GPU mps model then a subsequent forward using onnx will not match a forward in PyTorch.  If the same model is on the CPU or on CUDA then the export will produce a model that matches PyTorch.  Note that PyTorch forwards on MPS and on CPU match in this case.

The example below shows my best effort at a minimum case.  There needs to be more than one feature on the BatchNorm2d.  You have to perform one forward before using eval() on the model so that the BatchNorm does a computation.

import copy
import torch
import torch
from torch import nn
import numpy as np
import os, sys
import onnx
import onnxruntime

y = torch.randn(1, 2, 3, 6)
device = 'cuda'
ygpu = y.to(device=device)
#conv = nn.Conv2d(in_channels=57, out_channels=1, kernel_size=(1,3), stride=2)
gpuconv = nn.Sequential(
            nn.Conv2d(2, 2, stride=1, kernel_size=3),
            nn.BatchNorm2d(num_features=2)
            ).to(device=device)
xgpu = gpuconv(ygpu)
gpuconv.eval()
xgpu = gpuconv(ygpu)
conv = copy.deepcopy(gpuconv).to(device='cpu')
x = conv(y)

print(torch.allclose(x, xgpu.to(device='cpu'),  rtol=1e-03, atol=1e-05))
torch.onnx.export(conv, y, 'test.onnx', export_params=True)
torch.onnx.export(gpuconv, ygpu, 'testgpu.onnx', export_params=True)
onnx_model = onnx.load('testgpu.onnx')
onnx.checker.check_model(onnx_model)

ort_session = onnxruntime.InferenceSession('testgpu.onnx')

def to_numpy(tensor):
    return np.array(tensor.to('cpu').detach().numpy() if tensor.requires_grad else tensor.cpu().numpy())

# compute ONNX Runtime output prediction

ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(y)}
ort_outs = np.array(ort_session.run(None, ort_inputs)).squeeze()
torch_out_numpy = to_numpy(x).squeeze()

# compare ONNX Runtime and PyTorch results
print(np.allclose(torch_out_numpy,ort_outs, rtol=1e-03, atol=1e-05))

**Comment by jcwchen** [comment]
Hi @NicolasNierenbergPersyst,
IIUC, you are saying the same PyTorch code will produce different ONNX model on different machines? (M1 GPU v.s. others) If so, it seems like a question for the PyTorch-ONNX exporter (torch.onnx.export). Please raise this issue in PyTorch repo (converter team is there) and tag ONNX to get better assistance from the converter experts.

Another possibility is the converted ONNX models are basically the same, but onnxruntime produces different inference results on different machines. Then this issue might exist in onnxruntime instead.

**Comment by NicolasNierenbergPersyst** [comment]
I’m saying exactly what the code says.  On my Mac the gpu and cpu produce the same result on PyTorch but the onnx model produces a different result if it is exported from a GPU Model than a CPU model.  This doesn’t happen on COLAB where onnx produces the same result when exported by either a cpu or cuda model.  

I don’t know what would cause that but this is a simple code example that demonstrates the issue.

**Comment by jcwchen** [comment]
I don't have M1 GPU to repro this issue thru your code so it would be very helpful if you can provide `testgpu.onnx` and `test.onnx`. Then I can take a closer look at these 2 models and identify it's an issue in PyTorch-ONNX exporter or ONNX Runtime. Thanks!

**Comment by NicolasNierenbergPersyst** [comment]
[Archive.zip](https://github.com/onnx/onnx/files/9311347/Archive.zip)
Here you go.

**Comment by jcwchen** [comment]
Thanks for providing the models -- Now I am more confident that it should be a question for the PyTorch-ONNX exporter. Although the graph looks exactly the same, the initializer values they have are slightly different. I am not sure whether it is expected when exporting ONNX models on different machines, but it looks suspicious. Besides, you are using onnxruntime instead of onnxruntime-gpu, am I correct? If yes, onnxruntime should always use CPU ep by default so this issue is more likely to be caused by the exporter.

**Comment by NicolasNierenbergPersyst** [comment]
@jcwchen 

just to be clear as shown above both exports are on the same machine.  Just that in one case the model and the data are on the GPU and in the other case on the CPU.

Also it doesn’t matter to me but it seems that the onnx exporter  on CUDA works as expected.  I don’t think it is using CPU.

**Comment by justinchuby** [comment]
https://github.com/pytorch/pytorch/issues/83230

**Comment by jcwchen** [comment]
This issue seems resolved from the PyTorch exporter so I will close this one now. Feel free to reopen if there is still an issue. Thanks.
